sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,727 @@
+import math
+from functools import lru_cache
+
+import numba_dpex as dpex
+import dpctl
+
```

Comment:
Maybe add a top level comment to this module to explain that we use factory function to define kernel-wide constants at runtime so as to allow the llvm compiler to generate optimized code that can safely assume some dimensions of the problem to be constant and therefore allow loop unrolling or similar.

Reply:
Thanks for writing the comment bellow based on this suggestion, @fcharras.